libnd4j/include/templatemath.h

```diff
@@ -1575,27 +1622,143 @@ inline __device__ Nd4jLong nd4j_atomicMul<Nd4jLong>(Nd4jLong* address, Nd4jLong
 
 template <>
 inline __device__ bfloat16 nd4j_atomicMul<bfloat16>(bfloat16* address, bfloat16 val) {
-	return internal_16bit_atomicMul<bfloat16>(address, val);
+//	return internal_16bit_atomicMul<bfloat16>(address, val);
+	auto address_as_ull = (int*) address;
+
+	auto addr = (long)(address);
+	bool misaligned = addr & 0x3;
+
+	if (misaligned)
+		address_as_ull = (int*) (address - 1);
+
+	BPAIR old, assumed, fresh;
+
+	old.W = *address_as_ull;
+	do {
+
+		if (!misaligned) {
+			bfloat16 res = old.B.H * val;
+			fresh.B.H = res;
+			fresh.B.L = old.B.L;
+		} else {
+			bfloat16 res = old.B.L * val;
+			fresh.B.L = res;
+			fresh.B.H = old.B.H;
+		}
+
+		assumed.W = old.W;
+		old.W = atomicCAS(address_as_ull, assumed.W, fresh.W);
+	} while (assumed.W != old.W);
+
+	if (!misaligned) return old.B.H;
+	else return old.B.L;
 }
 
 template <>
 inline __device__ float16 nd4j_atomicMul<float16>(float16* address, float16 val) {
-	return internal_16bit_atomicMul<float16>(address, val);
+//	return internal_16bit_atomicMul<float16>(address, val);
```

Comment:
Why? We've found what was wrong with mul. So get internal_16bit_atomicMul back then?